src/CoreWCF.NetTcp/tests/TransportWithMessageCredentialNetTCPTest.cs

```diff
@@ -0,0 +1,139 @@
+                finally
+                {
+                    ((IChannel)channel).Abort();
+                    factory.Close();
+                }
+            }
+        }
+
+        private static Random random = new Random();
+        private static string RandomString(int length)
+        {
+            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+            return new string(Enumerable.Repeat(chars, length)
+              .Select(s => s[random.Next(s.Length)]).ToArray());
+        }
+
+        public class StartUpPermissionBaseForTC
+        {
+
+            public const string WindowsAuthRelativePath = "/nettcp.svc/windows-auth";
+            public const string NoSecurityRelativePath = "/nettcp.svc/security-none";
+            public void ConfigureServices(IServiceCollection services)
+            {
+                services.AddServiceModelServices();
+            }
+
+            public void ChangeHostBehavior(ServiceHostBase host)
+            {
+                var srvCredentials = new CoreWCF.Description.ServiceCredentials();
+                srvCredentials.ServiceCertificate.Certificate=  ServiceHelper.GetServiceCertificate();
+                srvCredentials.UserNameAuthentication.UserNamePasswordValidationMode
+                = CoreWCF.Security.UserNamePasswordValidationMode.Custom;
+                srvCredentials.UserNameAuthentication.CustomUserNamePasswordValidator
+                    = new CustomTestValidator();
+                host.Description.Behaviors.Add(srvCredentials);
+            }
+
+            public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+            {
+                CoreWCF.NetTcpBinding serverBinding = new CoreWCF.NetTcpBinding(SecurityMode.TransportWithMessageCredential);
+                serverBinding.Security.Message.ClientCredentialType = MessageCredentialType.UserName;
+                app.UseServiceModel(builder =>
+                {
+                    builder.AddService<Services.TestService>();
+                    builder.AddServiceEndpoint<Services.TestService, ServiceContract.ITestService>(serverBinding, WindowsAuthRelativePath);
+                    Action<ServiceHostBase> serviceHost = host => ChangeHostBehavior(host);
+                    builder.ConfigureServiceHostBase<Services.TestService>(serviceHost);
+                });
+            }
+
+            internal class CustomTestValidator : UserNamePasswordValidator
+            {
+                public override void Validate(string userName, string password)
+                {
+                    if (string.Compare(userName, "testuser@corewcf", StringComparison.OrdinalIgnoreCase) == 0 && password.Length >0)
+                    {
+                        //Write custom logic
+                        return;
+                    }
+                    else
+                    {
+                        throw new CommunicationException("Permission Denied");
+                    }
```

Comment:
Any reason you changed the type of exception thrown here? Any exception should work to be thrown, so if it doesn't, that's a problem.